File: src/main/java/gregtechlite/gtlitecore/mixins/hooks/Implemented.java
package gregtechlite.gtlitecore.mixins.hooks;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * Mark a mixins is an adhoc resolve which is implemented from upstream repositories or
 * unrelease contents in our mods, or just referenced some external contents.
 */
@Target(ElementType.TYPE)
@Retention(RetentionPolicy.SOURCE)
public @interface Implemented
{
    /**
     * A source for the implemented contents.
     * <p>
     * Here are some allowed source format:
     * <ul>
     *     <li>URL format link, e.g. GitHub Issue/Pull Request link</li>
     *     <li>Reference with correct format, like paper reference format</li>
     *     <li>Blog link with its author name (or id)</li>
     * </ul>
     *
     * @return The extract source for implemented contents.
     */
    String[] at() default "";
}

File: src/main/kotlin/gregtechlite/gtlitecore/mixins/hooks/Implemented.kt
package gregtechlite.gtlitecore.mixins.hooks

/**
 * Mark a mixins is an adhoc resolve which is implemented from upstream repositories or
 * unrelease contents in our mods, or just referenced some external contents.
 *
 * @param at A source for the implemented contents. Here are some allowed source format:
 *           1. URL format link, e.g. GitHub Issue/Pull Request link.
 *           2. Reference with correct format, like paper reference format.
 *           3. Blog link with its author name (or id)
 */
@Target(AnnotationTarget.CLASS)
@Retention(AnnotationRetention.SOURCE)
annotation class Implemented(val at: Array<String> = [""])